lib/hanami/action/application_action.rb

```diff
@@ -101,11 +101,15 @@ def view_context_options(req, res)
           {request: req, response: res}
         end
 
-        # Automatically render the view, if the body hasn't been populated yet
         def finish(req, res, halted)
-          res.render(view, **req.params, **res.exposures) if view && res.body.empty?
+          res.render(view, **req.params, **res.exposures) if render?(res)
           super
         end
+
+        # This can be overridden to enable/disable automatic rendering
```

Comment:
```suggestion
        # Decide if render the current response with the associated view.
        # This can be overridden to enable/disable automatic rendering.
        #
        # @param [Hanami::Action::Response]
        # @return [TrueClass,FalseClass]
        #
        # @since 2.0.0
        # @api public
```